### DISCUSSION TITLE
PaddleOCR OCR analyzes Left-to-Right instead of Right-to-Left for Arabic- How to process RTL languages correctly?

### DISCUSSION THREAD

**Discussion by roostaeecs** [PaddlePaddle/PaddleOCR#14971] [Q&A]
### 🔎 Search before asking

- [x] I have searched the PaddleOCR [Docs](https://paddlepaddle.github.io/PaddleOCR/) and found no similar bug report.
- [x] I have searched the PaddleOCR [Issues](https://github.com/PaddlePaddle/PaddleOCR/issues) and found no similar bug report.
- [x] I have searched the PaddleOCR [Discussions](https://github.com/PaddlePaddle/PaddleOCR/discussions) and found no similar bug report.

### 🐛 Bug (问题描述)

I am using PaddleOCR with the Arabic language model (lang='ar') to perform OCR on Arabic images. While PaddleOCR correctly recognizes the Arabic characters, it processes the text in a Left-to-Right (LTR) order, which is incorrect for Arabic, a Right-to-Left (RTL) language. This results in the words and sentences being in reverse order.

I have reviewed the paddleocr --help output to see if there are any options to explicitly set the text direction or handle RTL languages like Arabic.

My question is:

Is there a specific option in PaddleOCR, possibly using ocr_order_method or another parameter, to correctly handle Right-to-Left languages like Arabic and ensure the output text is in the correct RTL order?

If there isn't a built-in option, what are the recommended workarounds to post-process the OCR output to reorder the text correctly for RTL languages in Python?

Any guidance or solutions on how to get PaddleOCR to output Arabic text in the correct Right-to-Left order would be greatly appreciated.

### 🏃‍♂️ Environment (运行环境)

OS Windows 11
Environment Pycharm
Python 3.8
PaddOCR 2.10

### 🌰 Minimal Reproducible Example (最小可复现问题的Demo)

from paddleocr import PaddleOCR, draw_ocr

ocr = PaddleOCR(use_angle_cls=True, lang='Ar') 
img_path = 'image5.jpg'
result = ocr.ocr(img_path, cls=True)
for idx in range(len(result)):
    res = result[idx]
    for line in res:
        print(line)

# draw result
from PIL import Image
result = result[0]
image = Image.open(img_path).convert('RGB')
boxes = [line[0] for line in result]
txts = [line[1][0] for line in result]
scores = [line[1][1] for line in result]
im_show = draw_ocr(image, boxes, txts, scores, font_path='./doc/fonts/arabic.ttf')
im_show = Image.fromarray(im_show)
im_show.save('result.jpg')

**Comment by GreatV** [comment]
PaddleOCR currently does not fully support automatic handling of Right-to-Left (RTL) languages like Arabic in terms of reading order. While the model can accurately recognize Arabic characters, the OCR output is returned in a Left-to-Right (LTR) sequence, which leads to reversed word and sentence structures.

As of PaddleOCR v2.10, there is no built-in parameter such as ocr_order_method or an automatic RTL layout analyzer that handles RTL script flow natively. This is a known limitation for languages like Arabic, Persian, and Hebrew.

📌 Workaround: Post-process the output to reverse the order of text

You can manually adjust the text direction using Python after obtaining the OCR results. Here's a suggested approach:

1. After running OCR, group detected lines or blocks based on their vertical position (y-coordinate).
2. Within each line/block, sort results from right to left instead of left to right.
3. Concatenate the texts in reverse order to form the correct RTL string.

Here is an example post-processing snippet:

```python
# Example post-processing for RTL (Arabic)
from paddleocr import PaddleOCR
from operator import itemgetter

ocr = PaddleOCR(use_angle_cls=True, lang='ar')
img_path = 'image5.jpg'
result = ocr.ocr(img_path, cls=True)[0]

# Group lines based on y-coordinate (with some threshold)
lines = {}
threshold = 10

for box in result:
    position = box[0]
    text = box[1][0]
    y_center = (position[0][1] + position[2][1]) / 2
    added = False
    for key in lines:
        if abs(y_center - key) < threshold:
            lines[key].append((position, text))
            added = True
            break
    if not added:
        lines[y_center] = [(position, text)]

# Sort lines top to bottom
sorted_lines = sorted(lines.items(), key=lambda x: x[0])

for y, line_items in sorted_lines:
    # Sort boxes right to left based on x-coordinate of the rightmost point
    sorted_line = sorted(line_items, key=lambda x: -max(p[0] for p in x[0]))
    rtl_line = ' '.join([text for _, text in sorted_line])
    print(rtl_line)
```

📝 Notes:

- The sorting strategy above assumes relatively horizontal text lines. For more complex layouts (e.g., curved or multi-column text), further layout analysis may be needed.
- You may use libraries like python-bidi or arabic-reshaper along with an appropriate Arabic font to render visually correct RTL text, especially when displaying or printing.

🔮 Future Support:

While PaddleOCR supports multilingual OCR including Arabic in recognition, it currently lacks native document layout analysis or direction correction for RTL scripts. It’s recommended to follow or raise a feature request in the PaddleOCR GitHub Issues for improved RTL language support.

No currently available Issues or Discussions reference this particular RTL text order feature in PaddleOCR directly. You may consider opening a new Feature Request on GitHub.

---
*Response generated by 🤖 [feifei-bot](https://github.com/GreatV/feifei) | chatgpt-4o-latest*

 ⚠️ This is an AI-generated response. Please verify any critical information.

**Comment by cqray1990** [comment]
how to annotated for real data， reading order is from right to left,the cut images for training is not suitable for the label
@GreatV @roostaeecs

**Comment by RivoksLab** [comment]
**Two separate problems get conflated here, and only one of them is ordering**

We spent a while on RTL with PaddleOCR (Hebrew rather than Arabic, but the
mechanism should be identical) and it took us too long to realise these are
distinct failures:

**1. Ordering — a rendering/storage problem, not a model problem.**
The model output is already in *logical* order (the order you read the
characters). What breaks is applying `python-bidi`'s `get_display()` too early.
It is for rendering only — a terminal, a PDF, a GUI. Apply it before storage,
training or CER scoring and you silently corrupt everything downstream while
every metric still looks fine. Applying it once by mistake cost us 114,755
corrupted annotations.

A cheap way to audit a dataset someone hands you: measure the fraction of RTL
words whose first character is a final-form letter. Correct logical-order data
sits at 1–6%; a visual-order pool shows 20–27%.

**2. Deletion — this one IS the model, and it is the harder problem.**
When a run of Latin letters or digits is embedded *inside* RTL text (an
institution name, a date, a catalogue number, a citation), a CTC decoder tends
to **blank-collapse the entire run**. Not misread — deleted. You get back clean,
fluent, confident RTL text with a hole in it, and no error signal anywhere.

The diagnostic is where it *doesn't* happen. On our Hebrew models: isolated
Latin 1.20% CER, isolated digits 1.41%, pure Hebrew 0.64%. Every glyph class is
solved in isolation. Only the mixed boundary fails — Hebrew+Latin 12.63%. It
also isn't an ordering bug; the pure-reorder rate is near zero. Our read is that
CTC's monotonic alignment and the actual image alignment disagree at the
direction flip, and blanks are the decoder's cheapest way out.

**The fix that worked for us:** keep the backbone, change the decoding head. An
attention decoder (NRTR) has no monotonic constraint and reads the island
natively. Same SVTRv2 backbone, head swapped:

| slice | CTC | attention |
|---|---:|---:|
| Hebrew + embedded Latin (n=233 real crops) | 12.63% | **2.33%** |
| Hebrew + digits | 4.99% | **0.56%** |

Attention is slower, so we run a **script-gated cascade**: CTC first on
everything, fall back to attention only when the CTC output contains a
Latin/digit character or comes back low-confidence. About 10% of crops in
practice, so you get the attention ceiling at roughly 1.25–1.36× CTC latency,
no retraining.

**If you are benchmarking RTL, measure the mixed-script slice separately.** A
single aggregate CER hides this completely.

We have only tested this on Hebrew. If it reproduces on Arabic, Farsi or Urdu we
would genuinely like to know — please say so if you try it.

Numbers are in-house on our own test sets, vendor-reported and not
independently verified. Full writeup, including the negative results:
https://github.com/RivoksLab/paddleocr-hebrew/blob/main/docs/FINDINGS.md
Models (Apache-2.0, ONNX, finetuned from PaddleOCR):
https://huggingface.co/Rivok/paddleocr-hebrew
